File: pwn.college/helpers/example-arith-tcp.c
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <netdb.h>
#include <sys/socket.h>

int pwncollege() {
    return 0;
}

int create_fifo(char *pipe, char *color) {
    if (access(pipe, F_OK) == 0) {
        printf("%sNamed pipe %s already exists.\n", color, pipe);
    } else {
        printf("%sCreate named pipe %s using mknod.\n", color, pipe);
        syscall(133, pipe, 0010000 | 0644, 0);
    }
}

int main(int argc, char *argv[]) {
    char *colors[5] = {"\033[0;32m", "\033[0;33m", "\033[0;34m", "\033[0;35m", "\033[0;0m"};
    pid_t pid;
    int npfd, i, soutfd;
    FILE *outf;
    //{"/challenge/embryoio_level142", NULL, NULL, NULL},
    char *cmds[3][4] = {{"/usr/bin/python", "arithmetich.py", colors[0], NULL}, {"", NULL}, {"", NULL}};
    int sockfd;
    int len = 1;
    struct sockaddr_in servaddr;
    char buf[1024];

    create_fifo("pipe1", colors[4]);
    create_fifo("pipe2", colors[4]);

    pid = syscall(39);
    printf("%sParent %i.\n", colors[4], (int)pid);

    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        perror("Open socket error. Exit child.");
        syscall(60, 1);
    }

    bzero(&servaddr, sizeof(servaddr));
    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    servaddr.sin_port = htons(1493);

    if (connect(sockfd, (struct sockaddr*)&servaddr, sizeof(servaddr)) != 0) {
        perror("Connect error. Exit child.");
        syscall(60, 1);
    }

    for (i=0; i<3; ++i) {
        printf("%sFork.\n", colors[4]);
        pid = syscall(57);

        if (pid == -1) {
            perror("Fork failed. Exit child.");
            syscall(60, 1);
        }
        else if (pid == 0) {
            pid = syscall(39);
            printf("%sChild  %i %s.\n", colors[i], (int)pid, cmds[i][0]);

            // We do this to prevent any buffered output from messing with challenge stdin
            // after chaining childs stdins and outs.
            printf("%sDup stdout.\n", colors[i]);
            soutfd = syscall(32,1);
            outf = fdopen(soutfd, "w");

            if (*cmds[i][0] == '\0' && i == 1) {
                fprintf(outf, "%sOpen pipe2.\n", colors[i]);
                npfd = syscall(2, "pipe2", 01, 0666);

                while ((len=read(sockfd, buf, 1024)) > 0) {
                    write(npfd, buf, len);
                }

                fprintf(outf, "%sClose socket.\n", colors[i]);
                syscall(3, sockfd);
                fprintf(outf, "%sClose pipe2.\n", colors[i]);
                syscall(3, npfd);

                fprintf(outf, "%sExit child.\n", colors[i]);
                syscall(60, 1);
            }

            if (*cmds[i][0] == '\0' && i == 2) {
                fprintf(outf, "%sOpen pipe1.\n", colors[i]);
                npfd = syscall(2, "pipe1", 00, 0666);

                while ((len=read(npfd, buf, 1024)) > 0) {
                    write(sockfd, buf, len);
                }

                fprintf(outf, "%sClose socket.\n", colors[i]);
                syscall(3, sockfd);
                fprintf(outf, "%sClose pipe1.\n", colors[i]);
                syscall(3, npfd);

                fprintf(outf, "%sExit child.\n", colors[i]);
                syscall(60, 1);
            }

            //Execve args and env arrays must be NULL terminated
            fprintf(outf, "%sExec %s.\n", colors[i], cmds[i][0]);
            fprintf(outf, "%sClosing stdout file before exec.\n", colors[i]);
            syscall(3, soutfd);
            syscall(59, cmds[i][0], cmds[i], NULL);
            perror("Execve error. Exit child.");
            syscall(60, 1);
        }
    }

    while ((pid = syscall(61,  -1, NULL, 0, NULL)) > 0) {
        printf("%sChild %i terminated.\n", colors[4], (int)pid);
    }

    return 0;
}

File: pwn.college/helpers/example-arith-tcp.py
import multiprocessing as mp
import os
import pathlib
import socket
import subprocess


def create_fifo(pipe_path):
  if(pathlib.Path(pipe_path).exists()):
    print(f"Named pipe {pipe_path} already exists.")
    return
  try:
    os.mkfifo(pipe_path)
    print(f"Named pipe created at {pipe_path}.")
  except OSError as e:
    print(f"Error: {e}.")


def child0(s, line):
  print("Child0.")
  with open("pipe2", "w") as outf:
    f = s.makefile()

    while(line):
      line = f.readline()
      #print(f"Write line: {line}", end="")
      outf.write(line)
      outf.flush()


def child1(s, line):
  print("Child1.")
  with open("pipe1", "r") as inf:
    f = s.makefile("w")

    while(line):
      line = inf.readline()
      #print(f"Read line: {line}")
      f.write(line)
      f.flush()


sp = subprocess.Popen(["/challenge/embryoio_level141"], stderr=subprocess.STDOUT, stdout=subprocess.DEVNULL)

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect(("127.0.0.1", 1892))

create_fifo("pipe1")
create_fifo("pipe2")

p0 = mp.Process(target=child0, args=(s, "line"))
p0.start()

p1 = mp.Process(target=child1, args=(s, "line"))
p1.start()

subprocess.run(["python", "arithmetich.py", "\033[0;32m"])

s.close()
print("\033[0;0m", end="")
p0.join()
p1.join()
sp.kill()
